src/lib/clients.rs

```diff
@@ -262,71 +252,52 @@ impl StateClient for State {
 
     /// Find notes for a view key, starting from the given block height.
     fn fetch_notes(&self, vk: &ViewKey) -> Result<Vec<Note>, Self::Error> {
-        prompt::status("Fetching block height...");
-        let psk = &vk.public_spend_key().to_bytes()[..];
-        prompt::status("Fetching cached notes...");
-        let cached_block_height = self.cache.last_block_height(psk);
-        let cached_notes = self.cache.cached_notes(psk)?;
-
-        let msg = GetNotesOwnedByRequest {
-            height: cached_block_height,
-            vk: vk.to_bytes().to_vec(),
+        let mut state = self.inner.lock().unwrap();
+
+        prompt::status("Getting cached block height...");
+        let psk = vk.public_spend_key();
+        let last_height = state.cache.last_height(psk)?;
+
+        prompt::status("Fetching fresh notes...");
+        let msg = GetNotesRequest {
+            height: last_height,
+            vk: vec![], // empty vector means *all* notes will be streamed
         };
         let req = tonic::Request::new(msg);
+        let mut stream = state.client.get_notes(req).wait()?.into_inner();
+        prompt::status("Connection established...");
 
-        prompt::status("Fetching fresh notes...");
-        let mut state = self.client.lock().unwrap();
-        let res = block_in_place(move || {
-            Handle::current()
-                .block_on(async move { state.get_notes_owned_by(req).await })
-        })?
-        .into_inner();
-
-        prompt::status("Notes received!");
-        prompt::status("Handling notes...");
-
-        // collect notes
-        let mut fresh_notes: Vec<Note> = res
-            .notes
-            .into_iter()
-            .flat_map(|n| {
-                let mut bytes = [0u8; Note::SIZE];
-                bytes.copy_from_slice(&n);
-
-                let note = Note::from_bytes(&bytes).unwrap();
-                let key = note.hash().to_bytes().to_vec();
-                match cached_notes.contains_key(&key) {
-                    true => None,
-                    false => Some(note),
-                }
-            })
-            .collect();
+        prompt::status("Streaming notes...");
 
-        if !fresh_notes.is_empty() {
-            prompt::status("Caching notes...");
-            self.cache.persist_notes(psk, &fresh_notes[..])?;
-            self.cache.persist_block_height(psk, res.height)?;
-            prompt::status("Cache updated!");
+        while let Some(item) = stream.next().wait() {
+            let rsp = item?;
+            let note = Note::from_slice(&rsp.note)?;
+
+            if vk.owns(&note) {
+                state.cache.insert(psk, note, rsp.height)?;
```

Comment:
Does this mean that I always fetch from the beginning if I own no notes?

Maybe it worth to set the key_data.last_height

Reply:
I feel this makes more sense as a new issue than as a review here, since the previous cache didn't have this feature either. I'll create an issue around it.

Reply:
https://github.com/dusk-network/wallet-cli/issues/36